torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -3337,10 +3337,14 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
       fn(tensor, comm_, ncclStream, p2pTargetRank),
       ncclComm->getNcclCommFailureReason());
 #else
-  C10D_NCCL_CHECK_TIMEOUT(
-      fn(tensor, comm_, ncclStream, p2pTargetRank),
-      ncclComm->getNcclComm(),
-      ncclComm->getNcclCommFailureReason());
+  // In non-blocking mode, we need to use ncclGroup semantics to ensure that the
+  // kernel is enqueued for single-P2P ops.  Otherwise, the event record below
```

Comment:
i don't understand this explanation.  
 - why is this specific to p2p ops?  when in non-blocking mode, do all collectives need to be in a 'group' to ensure they are correctly enqueued before the subsequent event record?

Reply:
I feel we would need better contract / documentation from NCCL in terms of what `ncclSuccess` means for P2P calls -- i.e. whether kernel is guaranteed to have been enqueued or not. In this instant case, it seems not, so a group call is used to help it.

Reply:
Re why specific to P2P ops:
P2P ops involves dynamic connection establishment. This is the part made non-blocking in non-blocking mode.

Reply:
hmm, ok that kinda makes sense. 

however i have some concerns still: this change would effectively turn all P2P ops into grouped-P2P ops of size 1.  Does grouping add overhead to P2P performance that could be avoided in the blocking case?  (Overhead either on CPU side or on GPU side- does a 'group' p2p run a different kernel for example?

Reply:
No difference on GPU side.
Re CPU side, I wouldn't say there are not more instructions, but the overhead is minimal.